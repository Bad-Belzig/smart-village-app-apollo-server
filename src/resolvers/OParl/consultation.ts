import { FilterQuery } from 'mongoose';

import {
  AgendaItem,
  Consultation,
  IConsultation,
  IConsultationSchema,
  Meeting,
  Organization,
  Paper,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const consultationResolvers = {
  Query: {
    oParlConsultations: (
      _,
      args: { externalIds?: string[]; keyword?: string[] },
    ) => {
      const filter: FilterQuery<IConsultationSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return Consultation.find(filter);
    },
  },
  OParlConsultation: {
    paper: (parent: IConsultation) =>
      Paper.findOne({ externalId: parent.paper }),
    agendaItem: (parent: IConsultation) =>
      AgendaItem.findOne({ externalId: parent.agendaItem }),
    meeting: (parent: IConsultation) =>
      Meeting.findOne({ externalId: parent.meeting }),
    organization: async (
      parent: IConsultation,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Organization,
        parent.organization,
        args.offset,
        args.pageSize,
      ),
  },
};
