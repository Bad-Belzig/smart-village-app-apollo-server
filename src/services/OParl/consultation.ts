import { Consultation } from '../../models';
import { parseConsultation } from '../../parser';
import { getJson, updateOrCreateEntry } from './importHelpers';
import { ImportQueue, ImportQueueEntry, ImportType } from './ImportTypes';

export const importConsultation = async (
  value: unknown,
  queue?: ImportQueue,
) => {
  const json = await getJson(value);

  if (!json) return;

  const addToQueue: [ImportQueueEntry | ImportQueueEntry[], ImportType][] = [];

  if (json.agendaItem) {
    addToQueue.push([json.agendaItem, ImportType.AgendaItem]);
  }

  if (json.meeting) {
    addToQueue.push([json.meeting, ImportType.Meeting]);
  }

  if (json.organization) {
    addToQueue.push([json.organization, ImportType.Organization]);
  }

  if (json.paper) {
    addToQueue.push([json.paper, ImportType.Paper]);
  }

  try {
    return await updateOrCreateEntry(
      json,
      parseConsultation,
      Consultation,
      addToQueue,
      queue,
    );
  } catch (e) {
    console.log(`Error while importing consultation from ${value}:`, e);
  }
};
