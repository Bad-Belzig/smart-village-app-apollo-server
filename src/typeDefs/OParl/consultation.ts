import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const consultationGQL = gql`
  extend type Query {
    oParlConsultations(
      externalIds: [String!]
      keyword: [String!]
    ): [OParlConsultation!]
  }

  type OParlConsultation {
    agendaItem: OParlAgendaItem
    authoritative: Boolean
    meeting: OParlMeeting
    organization(offset: Int, pageSize: Int): [OParlOrganization!]
    paper: OParlPaper
    role: String
    ${oParlBase}
  }
`;
