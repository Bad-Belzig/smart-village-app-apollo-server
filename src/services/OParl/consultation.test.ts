import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../helpers';
import { importConsultation } from './consultation';

const helperFuncs = setupTestDataBase('test_consultation_import');

// TODO: change test url to something within our control, that uses all fields
const testUrl =
  'https://www.muenchen-transparent.de/oparl/v1.0/consultation/48823';
const testUrl2 =
  'https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/consultation/78';

describe('importing a consultation', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly for 1.0', basicImportTest(importConsultation, testUrl));

  it('imports properly for 1.1', basicImportTest(importConsultation, testUrl2));

  it(
    'import returns undefined on faulty url',
    invalidUrlTest(importConsultation),
  );

  afterAll(helperFuncs.afterAll);
});
