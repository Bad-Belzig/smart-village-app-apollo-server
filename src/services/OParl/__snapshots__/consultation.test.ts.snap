// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a consultation imports properly 1`] = `
Object {
  "__v": 0,
  "externalId": "https://www.muenchen-transparent.de/oparl/v1.0/consultation/48823",
  "keyword": Array [],
  "meeting": "https://www.muenchen-transparent.de/oparl/v1.0/meeting/5703",
  "organization": Array [
    "https://www.muenchen-transparent.de/oparl/v1.0/organization/gremium/5",
  ],
  "type": "https://schema.oparl.org/1.0/Consultation",
}
`;
